src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -12233,6 +12291,7 @@ private SyntaxList<SyntaxToken> ParseAnonymousFunctionModifiers()
 
         private LambdaExpressionSyntax ParseLambdaExpression()
         {
+            var attributes = ParseAttributeDeclarations(checkLambdaAttributesFeature: true);
```

Comment:
so this is technically a bit unclean, but *hopefully* not a problem.  THe main issue is that this makes attribute parsing context sensitive.  We generally like avoiding that as it makes incremental difficult.  Specifically, it means we couldn't jsut say "ah, we can reuse the attributes from a previous parse that are available in this location".

In your case, i think you're ok.  that's bacause you'll have one of the two cases:

1. the attribute decl is on a lambda before.  so it will have the feature diagnostic on it, so it will crumble if we try to reparse it.
2. the attribute decl is not on a lambda.  it is reused in ParseAttributeDecl, but we still add the diagnostic on afterwards.

This is subtle though and easy to get wrong.  in general, it can be preferable to just do the parse normally and have the caller (i.e. this method) do the check.

Reply:
> this makes attribute parsing context sensitive.  We generally like avoiding that as it makes incremental difficult.

I'm interested to understand how to handle the feature check in this method. Specifically, what would the resulting error be attached to that would simplify incremental parsing? (Perhaps I can address this in a subsequent PR.)

---
In reply to: [604500799](https://github.com/dotnet/roslyn/pull/52038#discussion_r604500799) [](ancestors = 604500799)